### PR TITLE
Global styles: background UI controls

### PR DESCRIPTION
**Pull request by ramonjd**
## Dev Note 📓 

https://github.com/WordPress/gutenberg/pull/59354#issuecomment-2138583222

## What?

Part of:

- https://github.com/WordPress/gutenberg/issues/54336

Follow up to:

- https://github.com/WordPress/gutenberg/pull/59354

Related:

- https://github.com/WordPress/gutenberg/pull/60100

This PR add background image UI controls to top-level global styles. 

### TODO

- Maybe E2E tests...

## Why?

Background styles can now be added to theme.json since https://github.com/WordPress/gutenberg/pull/59354.

The styles are currently only " top-level", which means they're added to `body` selector. 

This PR allows users to edit these styles in the site editor.

## How?

Similar to dimensions and typography panels, this PR refactors the background block supports component so that it can be shared between the site and post editors.

It moves all relevant tests and styles to the `/global-styles` directory.

## Testing Instructions

This is a big one, so get comfortable.

Firstly, check out this branch.

At this stage it's probably good to check if there are any regressions to background block supports. So, add a group block to a post, and make sure you can:

1. Add/remove/upload/reset a new background image (See https://github.com/WordPress/gutenberg/pull/53934 for test instructions)
2. Edit/reset the background position values (https://github.com/WordPress/gutenberg/pull/58592)
3. Edit/reset background size and repeat values (https://github.com/WordPress/gutenberg/pull/57005)

If that's all good, head over to the site editor.

I've been using empty theme to test. Here is some test theme.json, with a relative path to the updated theme.json schema:

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"background": {
			"backgroundImage": true,
			"backgroundSize": true
		}
	},
	"styles": {
		"color": {
			"background": "tomato"
		},
		"background": {
			"backgroundImage": {
				"url": "https://i0.wp.com/wordpress.org/files/2023/12/sotw-dotorg-drawer.png?w=1807&ssl=1",
				"source": "file"
			},
			"backgroundPosition": "30% 30%",
			"backgroundRepeat": "repeat",
			"backgroundSize": "cover"
		}
	}
}
```

Ensure that:

1. Any inherited values from theme.json are present.
2. You can interact with the background controls in a similar way to above, e.g., at the block level, and
3. Saving the styles affects the body styles of your site. 
4. Also check that the background size defaults are: "cover" for Group blocks, and `undefined` for site backgrounds. See table images below.

| Group block background controls | Top-level global styles background controls |
|--------|--------|
| Screenshot 2024-03-21 at 3 25 15 pm | Screenshot 2024-03-21 at 3 24 56 pm | 

### Testing Instructions for Keyboard

In the global styles panel, ensure you can navigate to the "Background" item in the list to activate the background styles panel. There should be no regressions to keyboard interaction with background controls.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/6458278/8b4274e5-f0a9-4ab0-b34d-dc8d2e200255

### CODE REVIEW

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+		inheritedValue?.background?.backgroundSize;
+	const repeatValue =
+		style?.background?.backgroundRepeat ||
+		inheritedValue?.background?.backgroundRepeat;
+	const imageValue =
+		style?.background?.backgroundImage?.url ||
+		inheritedValue?.background?.backgroundImage?.url;
+	const positionValue =
+		style?.background?.backgroundPosition ||
+		inheritedValue?.background?.backgroundPosition;
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:351]: Should we care about incoming, named values from theme json, e.g., `"center center"`, or "top left"` etc.

I remember @andrewserong might have been doing some conversion in another PR, maybe unrelated to background. 🙃

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 2]
```diff
@@ -0,0 +1,580 @@
+/**
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:1]: Most of the contents of this file have been ported over 1:1 from `hooks/background.js` with the modifications necessary to handle inherited values from theme.json/global styles.

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+			panelId={ panelId }
+		>
+			<BackgroundImageToolsPanelItem
+				onChange={ onChange }
+				panelId={ panelId }
+				isShownByDefault={ defaultControls.backgroundImage }
+				style={ value }
+				inheritedValue={ inheritedValue }
+			/>
+			{ shouldShowBackgroundSizeControls && (
```
Comment (ramonjd) [inline packages/block-editor/src/components/global-styles/background-panel.js:579]: I think this is right... allow this panel if theme.json (and incoming block supports flag) has  `background.backgroundSize` set to `true`

File: packages/block-editor/src/components/global-styles/background-panel.js (hunk 1 above)
Comment (andrewserong) [reply packages/block-editor/src/components/global-styles/background-panel.js:351]: Ah, yes, we were originally looking at it when adding in `backgroundPosition` support but I dropped it as it made things a bit more complex and we were getting closer to the Beta 1 feature freeze. It'd be a good thing to add back in, but it could be its own PR.

**Review by andrewserong — COMMENTED** [review]
Very cool, this is fun to play with! I've just added the "Needs Design Feedback" label, since it looks like there'll be a few things to figure out.

The main question for me in playing around with it is: what should the default size values be when someone first uploads or adds an image? Here's how it looks for me with a regular photo on a very tall template (it results in the image being very zoomed in):

image

Defaulting to Cover works nicely for the Group block, but in practice, it feels like at the root of the site, folks might rarely want to use either Cover or Contain, since it'll be pretty inconsistent depending on the height of the current page. So perhaps in the root of global styles, defaulting to Fixed might work better?

Also, in global styles, what do you think about displaying the size controls by default? It looked a bit empty to me in the initial state after adding an image:

image

**Comment by ramonjd** [comment]
Thank you for testing @andrewserong 🙇🏻 

> what should the default size values be when someone first uploads or adds an image?

Fixed and no-repeat could work as a default - that's `background-size: auto;` right? 

Agree that it would allow users to see the image they've added, regardless of size

Then folks could change the settings themselves. 

It might also work for the theme.json side of things - Gutenberg setting the backgroundSize to `auto` by default seems less opinionated than setting it to `cover`. I don't know why 😄 


> Also, in global styles, what do you think about displaying the size controls by default?

Can do. 👍🏻 I basically copied the post editor.

**Comment by andrewserong** [comment]
> Fixed and no-repeat could work as a default - that's background-size: auto; right?

I think so — worth having a play with and seeing how it feels. It'll be interesting to then think through once we've added an image, which is the first control we find we repeatedly reach for after that 🤔 

> It might also work for the theme.json side of things - Gutenberg setting the backgroundSize to auto by default seems less opinionated than setting it to cover. I don't know why 😄

Yeah, it does to me, too. It's funny how these things seem to become clearer once a PR is up. When interacting with a particular block in the post editor, `cover` feels natural to me because you can see the boundaries of the block right there so it looks right if it covers that area, but global styles feels quite different conceptually.

> Can do. 👍🏻 I basically copied the post editor.

Nice one!

**Comment by ramonjd** [comment]
> worth having a play with and seeing how it feels.

@andrewserong I started a PR to test it out:

- https://github.com/WordPress/gutenberg/pull/59889

👍🏻

**Comment by ramonjd** [comment]
Because this PR contains a necessary refactor to share the background controls across block and global styles, I think we should look to get it merged first before we progress to any [UI refactoring](https://github.com/WordPress/gutenberg/pull/60151) - that way both (block and global styles) will get the benefit of the UI refactor and we'll save ourselves a horrible rebase. 😄

**Comment by andrewserong** [comment]
> I think we should look to get it merged first before we progress to any https://github.com/WordPress/gutenberg/pull/60151

Sounds good to me!

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+
+export default function BackgroundPanel( {
+	as: Wrapper = BackgroundToolsPanel,
+	value,
+	onChange,
+	inheritedValue = value,
+	settings,
+	panelId,
+	defaultControls = DEFAULT_CONTROLS,
+	defaultControlValues = {},
```
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/background-panel.js]: Tiny nit: would `defaultValues` be clearer, or are the default values tied to the concept of controls? To put the question differently: if the `defaultValues` object is identical to the shape of the `styles.background` part of the style object, do we still need the word `Control` in this param?

**Review by andrewserong — COMMENTED** [review]
Nice, this is feeling close to me! The only real issue I encountered is that it looks like saving the background image values fails on a site where the user does not have `unfiltered_html` capabilities.

To quickly test, in a plugin or anywhere else in PHP, I added the following locally:

```
add_action( 'init', 'kses_init_filters' );
```

Here's a quick screengrab:

https://github.com/WordPress/gutenberg/assets/14988353/0dbb78c5-aec5-45ff-813e-169898a3c0a2

Could we have missed something in the Theme JSON PHP class to ensure the background image doesn't get sanitized away?

Other than that, this is looking pretty good to me! Is the next step to ping some designers for feedback on the added background panel?

One tiny other note: will we need to add `background` to the [`getGlobalStylesChanges`](https://github.com/WordPress/gutenberg/blob/b8d84dbbc210d71600f2fda33a94623e2bd1adf4/packages/block-editor/src/components/global-styles/get-global-styles-changes.js#L14) list so that changes to the background image get included in the list of changes when saving global styles or viewing revisions?

File: packages/block-editor/src/components/global-styles/background-panel.js (hunk 4 above)
Comment (ramonjd) [reply packages/block-editor/src/components/global-styles/background-panel.js]: > the default values tied to the concept of controls?

The naming was intentional but I'm not in love with it, so happy to update if `defaultValues` is clearer. I think the shorter form you suggested communicates enough as well. 👍🏻 

To explain the motivation, "default control value" is what the control state should be by default, independent of any style values.

The reason being, blocks have a default `backgroundSize` of `"cover"`, but the value is added to the block props dynamically and are not present in the block attribute style object. 

So to indicate that `"cover"` has indeed been applied to the block, the control needs to be told to indicate **Cover** even where there is no attribute style value.

**Comment by ramonjd** [comment]
Thanks for taking this PR for a spin @andrewserong 

> Could we have missed something in the Theme JSON PHP class to ensure the background image doesn't get sanitized away?

🤔 Good catch. I'll look into it.

> Is the next step to ping some designers for feedback on the added background panel?

Yeah, I think the only specific UX issue is where to access the top-level background styles in Global styles. I think "Background" in the list makes sense, given that it's its own block support and will, presumably, appear so in the block settings.

**Comment by ramonjd** [comment]
> Could we have missed something in the Theme JSON PHP class to ensure the background image doesn't get sanitized away?

Just jotting down a couple of notes.

Looking at the [wp_filter_global_styles_post](https://developer.wordpress.org/reference/functions/wp_filter_global_styles_post/) filter (used in `kses_init_filters`), the target search area is probably around `WP_Theme_JSON_Gutenberg::remove_insecure_properties` and `WP_Theme_JSON_Gutenberg::is_safe_css_declaration`.

Not sure yet, I'll continue testing later, but maybe because `background-repeat` isn't featured in `safecss_filter_attr` 🤔 

https://github.com/WordPress/wordpress-develop/blob/6.4/src/wp-includes/kses.php#L2332

**Comment by andrewserong** [comment]
> Not sure yet, I'll continue testing later, but maybe because background-repeat isn't featured in safecss_filter_attr 🤔

Background repeat was added in 6.5 ([this line](https://github.com/WordPress/wordpress-develop/blob/0e284447951869fd3ff58b345ec070b0ab97d9bc/src/wp-includes/kses.php#L2338)) and in Gutenberg, we've got it added [here](https://github.com/WordPress/gutenberg/blob/071082ce485b36f84ca654de5f1a4e1a57d7eb80/lib/compat/wordpress-6.5/kses.php#L15), so (in theory) that one shouldn't be a problem?

> the target search area is probably around WP_Theme_JSON_Gutenberg::remove_insecure_properties and WP_Theme_JSON_Gutenberg::is_safe_css_declaration

They could be a good clue. I wonder if we need some extra handling since `backgroundImage` uses a nested `url` value, so it might be failing here? A couple of ideas: 1) if we know the value will always be handled / sanitised by the style engine, perhaps we could skip checking it here, or 2) add an additional check so that it knows how to handle a value in this property, where the value is further nested / handled in a different way?

**Comment by tellthemachines** [comment]
Just testing this PR and noticed that "Repeat" gets toggled off when moving from "Fixed" to "Contain", and toggled on when moving the other way. Is this deliberate? Would it be better to just leave it as is when moving from one size to another?

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+		// For media selections originated from a file upload.
+		if (
+			( media.media_type &&
+				media.media_type !== IMAGE_BACKGROUND_TYPE ) ||
+			( ! media.media_type &&
+				media.type &&
+				media.type !== IMAGE_BACKGROUND_TYPE )
+		) {
+			onUploadError(
+				__( 'Only images can be used as a background image.' )
```
Comment (tellthemachines) [inline packages/block-editor/src/components/global-styles/background-panel.js:233]: Oh, are there plans to allow video at any point? 👀

File: packages/block-editor/src/components/global-styles/background-panel.js (hunk 3 above)
Comment (tellthemachines) [reply packages/block-editor/src/components/global-styles/background-panel.js:579]: Is it `true` by default? I can see the controls in the global styles UI without changing anything in theme.json.

**Review by tellthemachines — COMMENTED** [review]
This is such an awesome feature. I love how well it works for adding patterns (didn't have to change any setting, just added the image!):

Screenshot 2024-03-26 at 3 54 42 PM

One thing that feels a bit weird is how we're able to set background size/position without having a background image in place, but that happens also for block background images so it's unrelated to this PR. It's weird in that we can fiddle with the controls, nothing visible happens, but there are unsaved changes (and there's a higher likelihood of that happening when all the controls are visible, unlike for block backgrounds).

File: packages/block-editor/src/components/global-styles/background-panel.js (hunk 5 above)
Comment (andrewserong) [reply packages/block-editor/src/components/global-styles/background-panel.js:233]: Just jumping in here since I think I originally added this line back in the earlier component. Adding video is listed [in the tracking issue](https://github.com/WordPress/gutenberg/issues/54336) — I think it'd be great to do eventually, but it'd involve conditionally outputting different markup for the block, which would be challenging at the block level for the block support, and potentially not really do-able or desirable for global styles? Or we'd need special handling for outputting the video at the root of the site.

Still, it'd be a great feature, and there's another use case for conditionally rendering different markup (potentially) for optimized background images, tracked in https://github.com/WordPress/gutenberg/issues/59113, so it would be worth investigating at some stage, and/or opening an issue for it if we have ideas for it that we'd like to capture.

File: packages/block-editor/src/components/global-styles/background-panel.js (hunk 3 above)
Comment (ramonjd) [reply packages/block-editor/src/components/global-styles/background-panel.js:579]: Hmm, something fishy is going on. It's `true` by default, but I can't see from where 🙃

**Comment by ramonjd** [comment]
Thanks a lot for testing @tellthemachines and @andrewserong 

> I wonder if we need some extra handling since backgroundImage uses a nested url value, so it might be failing here? 

Turns out that was it 🎉 

I ended up adding `background-image` to `INDIRECT_PROPERTIES_METADATA`. I think it fits as the value is "indirectly" derived from the array values.

Seems to work anyway. 😄 

> One thing that feels a bit weird is how we're able to set background size/position without having a background image in place, but that happens also for block background images so it's unrelated to this PR. 

Yeah, I'm thinking maybe in a follow up we could try hiding it until a background image is added?

This is the direction I'm going for in 

- https://github.com/WordPress/gutenberg/pull/60151

But the design allows it to a certain degree.

> "Repeat" gets toggled off when moving from "Fixed" to "Contain", and toggled on when moving the other way. Is this deliberate? Would it be better to just leave it as is when moving from one size to another?

This was pulled over from trunk:

https://github.com/WordPress/gutenberg/blame/trunk/packages/block-editor/src/hooks/background.js#L493

@andrewserong will know for sure, but maybe it's because repeat is toggled on automatically when flipping to "Fixed"?

Here's how the toggle works without that condition:

https://github.com/WordPress/gutenberg/assets/6458278/1e56d605-da31-444a-a179-8506095a434d

File: packages/block-editor/src/components/global-styles/background-panel.js (hunk 3 above)
Comment (ramonjd) [reply packages/block-editor/src/components/global-styles/background-panel.js:579]: Oh face palm moment. 

I forgot I added `"backgroundImage"` and `"backgroundSize"` to [appearanceTools](https://github.com/WordPress/gutenberg/blob/85045e4fae34335b2d14ce259f934731b6998a7a/lib/class-wp-theme-json-gutenberg.php#L675)

**Comment by ramonjd** [comment]
I added @jameskoster to keep him in the loop. 

**TL;DR**

This PR adds a new item for background controls to top-level global styles:

Screenshot 2024-03-26 at 7 21 23 pm

Playing around with UI refresh of controls over in https://github.com/WordPress/gutenberg/pull/60151

File: lib/class-wp-theme-json-gutenberg.php [hunk 6]
```diff
@@ -286,26 +286,31 @@ class WP_Theme_JSON_Gutenberg {
 	 *
 	 * Indirect properties are not output directly by `compute_style_properties`,
 	 * but are used elsewhere in the processing of global styles. The indirect
-	 * property is used to validate whether or not a style value is allowed.
```
Comment (ramonjd) [inline lib/class-wp-theme-json-gutenberg.php:289]: Just a personal obsession of mine.

File: lib/theme.json [hunk 7]
```diff
@@ -4,6 +4,10 @@
 	"settings": {
 		"appearanceTools": false,
 		"useRootPaddingAwareAlignments": false,
+		"background": {
+			"backgroundImage": true,
+			"backgroundSize": true
+		},
```
Comment (ramonjd) [inline lib/theme.json]: This turns the feature on by default.

Do folks think it's okay? We could wait until the feature is better tested and stable. 

I just added this change to discuss.

**Review by ramonjd — COMMENTED** [review]
I think this PR is ready for another check.

I appreciate the feedback!

File: lib/class-wp-theme-json-gutenberg.php [hunk 8]
```diff
@@ -286,26 +286,31 @@ class WP_Theme_JSON_Gutenberg {
...
 			array( 'spacing', 'blockGap', 'top' ),
 		),
-		'max-width'  => array(
+		'max-width'        => array(
 			array( 'layout', 'contentSize' ),
 			array( 'layout', 'wideSize' ),
 		),
+		'background-image' => array(
+			array( 'background', 'backgroundImage', 'url' ),
+			array( 'background', 'backgroundImage', 'source' ),
```
Comment (ramonjd) [inline lib/class-wp-theme-json-gutenberg.php:312]: I suspect this line isn't required. I'll test in the morning. Just adding note to self.

**Comment by jameskoster** [comment]
Thanks for the ping.

> > One thing that feels a bit weird is how we're able to set background size/position without having a background image in place, but that happens also for block background images so it's unrelated to this PR.
> 
> Yeah, I'm thinking maybe in a follow up we could try hiding it until a background image is added?

Big +1 to this.

Immediately after selecting a background image the Size control seems to be unset, shouldn't it be set to `fixed` with a width based on the source?

The repeated 'Background' heading feel a bit duplicative here:

Screenshot 2024-03-26 at 11 04 36

Appreciate this may be tricky but could we eliminate one and move the chevron or ellipsis to create something like:

Screenshot 2024-03-26 at 11 09 42



---

Follow-up territory:

* It'd be nice to support `background-attachment`, so you can create a cover + fixed combo (a common pattern).
* Multiple background images, each with unique properties.
* 'Repeat' control spacing looks a little off, could take inspiration from the drop cap implementation.

**Comment by ramonjd** [comment]
Thanks for the feedback @jameskoster 

> It'd be nice to support background-attachment, so you can create a cover + fixed combo (a common pattern).

This is on the short-term TODO list 👍🏻 

> The repeated 'Background' heading feel a bit duplicative here:

I'll sneak this one into this PR I think, but if folks are okay with it I'll do the rest in follow ups. This PR is a monster already 😆

File: lib/theme.json (hunk 7 above)
Comment (andrewserong) [reply lib/theme.json]: My understanding was that background image controls should be guarded behind the `appearanceTools` flag, similar to most of the other visual controls (border, margin, padding, etc). That said, I can see how this might feel a bit more like a feature than a simple styling control, and would be useful to expose more widely. If we're thinking of changing it, I'd recommend exploring that in a separate PR as I've seen the discussions between default to on or off run on quite long for some features (most recently shadow support).

Comment (ramonjd) [reply lib/theme.json]: No worries. Good points - exactly what I was after 👍🏻 

I'll remove. Many new themes have "appearanceTools" turned on anyway.

**Comment by ramonjd** [comment]
> The repeated 'Background' heading feel a bit duplicative here:
> 
> > I'll sneak this one into this PR I think, but if folks are okay with it I'll do the rest in follow ups. This PR is a monster already 😆



So much for that being straight forward 😄 

The tools panel component doesn't show the header, including the ellipsis menus if there is no panel heading.

https://github.com/WordPress/gutenberg/blob/2d338de9b076f98b40f86f2a8e7c8c4b90899f1f/packages/components/src/tools-panel/tools-panel-header/component.tsx#L170-L170

If I comment out that condition, there's a bit of a gap:

Screenshot 2024-03-27 at 10 50 13 am

So I optimistically tried injecting the background image component into the tools panel header, but it's wrapper in an h2 and reused in the drop down:

Screenshot 2024-03-27 at 10 55 45 am

So I think it might need a bit of rethink in another PR.

----

> So I think it might need a bit of rethink in another PR.

In a follow up PR, we could try using "Image" as the header and the file name in the media selector:

| With no image added | Image file name |
|--------|--------|
| Screenshot 2024-03-27 at 10 59 08 am | Screenshot 2024-03-27 at 10 58 41 am |

File: lib/class-wp-theme-json-gutenberg.php (hunk 8 above)
Comment (ramonjd) [reply lib/class-wp-theme-json-gutenberg.php:312]: > I suspect this line isn't required.

It most definitely is required 😄  Without the source it won't pass the `_backgroundImage?.source === 'file'` test in the style engine: 

https://github.com/WordPress/gutenberg/blob/d9058876c14a0e5e1c04f9bc63c9ec2085f0d000/packages/style-engine/src/styles/background/index.ts#L13-L13

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+		return __( 'Image covers the space evenly.' );
+	}
+	if ( value === 'contain' ) {
+		return __( 'Image is contained without distortion.' );
+	}
+	return __( 'Specify a fixed width.' );
+}
+
+/**
+ * Coverts decimal x and y coords from FocalPointPicker to percentage-based values
```
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/background-panel.js]: This must have been my typo from earlier PRs 😆 

```suggestion
 * Converts decimal x and y coords from FocalPointPicker to percentage-based values
```

File: packages/block-editor/src/components/global-styles/background-panel.js [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+	}
+
+	const x = isNaN( value.x ) ? 0.5 : value.x;
+	const y = isNaN( value.y ) ? 0.5 : value.y;
+
+	return `${ x * 100 }% ${ y * 100 }%`;
+};
+
+/**
+ * Coverts backgroundPosition value to x and y coords for FocalPointPicker.
```
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/background-panel.js]:
```suggestion
 * Converts backgroundPosition value to x and y coords for FocalPointPicker.
```

**Review by andrewserong — APPROVED** [review]
This is testing nicely for me now, and while there's a few more tweaks we've each brought up as ideas for follow-ups, given that this PR already includes a refactor of the panels and components, I think this is in a good place to land it, since it could be easier to continue tweaking in smaller follow-ups. What do you think?

Tests well:

✅ UI at the block level behaves largely as on `trunk`, and defaults to `cover`
✅ New panel and controls in global styles for site wide background images is working correctly, and defaults to the equivalent of `auto` 

> "Repeat" gets toggled off when moving from "Fixed" to "Contain", and toggled on when moving the other way. Is this deliberate? Would it be better to just leave it as is when moving from one size to another? ... 
> @andrewserong will know for sure, but maybe it's because repeat is toggled on automatically when flipping to "Fixed"?

From memory, we introduced this in the original background sizes PR based on an assumption of which settings folks would like if they have Contain selected or Fixed selected, so that they don't need to toggle the repeat control. I.e. if using Contain, the assumption was that at the block level folks wouldn't be using repeat, but if using Fixed, the assumption was that they would be.

This is very much up for revisiting, especially if it feels confusing! My only recommendation here would be to revisit it in a separate PR since this one is already doing quite a bit.

This LGTM! ✨ 

In a follow-up, it'd also be good to add the background image changes to the revisions / changes list. Just adding this in from an earlier comment, in case it was lost in the backscroll:

> One tiny other note: will we need to add background to the [getGlobalStylesChanges](https://github.com/WordPress/gutenberg/blob/b8d84dbbc210d71600f2fda33a94623e2bd1adf4/packages/block-editor/src/components/global-styles/get-global-styles-changes.js#L14) list so that changes to the background image get included in the list of changes when saving global styles or viewing revisions?

**Comment by ramonjd** [comment]
Thanks again @andrewserong 

> In a follow-up, it'd also be good to add the background image changes to the revisions / changes list. Just adding this in from an earlier comment, in case it was lost in the backscroll:

Oh, thanks for the nudge. I'd added that to my list but hadn't acknowledged it in the comments 👍🏻 It's in the works!

**Comment by andrewserong** [comment]
> Oh, thanks for the nudge. I'd added that to my list but hadn't acknowledged it in the comments 👍🏻 It's in the works!

No worries! Thanks again for all the detailed work here, it's such an exciting feature 😀

**Comment by tellthemachines** [comment]
> This is very much up for revisiting, especially if it feels confusing! My only recommendation here would be to revisit it in a separate PR since this one is already doing quite a bit.

Cool cool yeah definitely let's not block this PR. Thanks for clarifying!

**Comment by ramonjd** [comment]
Thanks again, everyone!

Ongoing follow ups:

- https://github.com/WordPress/gutenberg/pull/60100
- https://github.com/WordPress/gutenberg/pull/60229
- https://github.com/WordPress/gutenberg/pull/60151

**Comment by jameskoster** [comment]
> So I think it might need a bit of rethink in another PR.

I figured that might be the case :D 

> In a follow up PR, we could try using "Image" as the header and the file name in the media selector:

Seems a simple solution :)

**Comment by jameskoster** [comment]
@ramonjd we might also consider naming this whole section "Background image" rather than just "Background", which may mislead users into believing they can set the background _color_ here.

**Comment by burnuser** [comment]
Great (and overdue) feature, which was missing in FSE Themes from start. (And existing in Classic Themes from start.)

But without the "Fixed background" option like in Cover block, the result looks ... strange.
No "Fixed background" option in Group block is bad enough, but for a site background, the fixed background image position is standard and really needed.

**Comment by ramonjd** [comment]
> No "Fixed background" option in Group block is bad enough, but for a site background, the fixed background image position is standard and really needed.

It's the first item on the TODO list here for WordPress 6.6:

- https://github.com/WordPress/gutenberg/issues/54336
